src/ooxml/java/org/apache/poi/xssf/streaming/SheetDataWriter.java

```diff
@@ -354,88 +355,54 @@ protected void outputQuotedString(String s) throws IOException {
+                case "&":
                     _out.write("&amp;");
                     break;
-                case '"':
-                    writeLastChars(_out, chars, last, counter);
-                    last = counter + 1;
+                case "\"":
                     _out.write("&quot;");
                     break;
                 // Special characters
-                case '\n':
-                    writeLastChars(_out, chars, last, counter);
+                case "\n":
                     _out.write("&#xa;");
-                    last = counter + 1;
                     break;
-                case '\r':
-                    writeLastChars(_out, chars, last, counter);
+                case "\r":
                     _out.write("&#xd;");
-                    last = counter + 1;
                     break;
-                case '\t':
-                    writeLastChars(_out, chars, last, counter);
+                case "\t":
                     _out.write("&#x9;");
-                    last = counter + 1;
                     break;
-                case 0xa0:
-                    writeLastChars(_out, chars, last, counter);
+                case "\u00A0":
                     _out.write("&#xa0;");
-                    last = counter + 1;
                     break;
                 default:
-                    // YK: XmlBeans silently replaces all ISO control characters ( < 32) with question marks.
-                    // the same rule applies to "not a character" symbols.
-                    if (replaceWithQuestionMark(c)) {
-                        writeLastChars(_out, chars, last, counter);
-                        _out.write('?');
-                        last = counter + 1;
-                    }
-                    else if (Character.isHighSurrogate(c) || Character.isLowSurrogate(c)) {
-                        writeLastChars(_out, chars, last, counter);
-                        _out.write(c);
-                        last = counter + 1;
-                    }
-                    else if (c > 127) {
-                        writeLastChars(_out, chars, last, counter);
-                        last = counter + 1;
-                        // If the character is outside of ascii, write the
-                        // numeric value.
-                        _out.write("&#");
-                        _out.write(String.valueOf((int) c));
-                        _out.write(";");
+                    if (codepoint.length() == 1) {
+                        char c = codepoint.charAt(0);
+                        // YK: XmlBeans silently replaces all ISO control characters ( < 32) with question marks.
+                        // the same rule applies to "not a character" symbols.
+                        if (replaceWithQuestionMark(c)) {
+                            _out.write('?');
+                        } else {
+                            _out.write(c);
+                        }
+                    } else {
```

Comment:
Would this write out `&#FFFF;`, `FFFF`, or \uFFFF?

Reply:
This writes the char unescaped.